ws-calculator/src/main/java/org/egov/wsCalculation/service/DemandService.java

```diff
@@ -0,0 +1,737 @@
+
+@Service
+@Slf4j
+public class DemandService {
+
+	@Autowired
+	private ServiceRequestRepository repository;
+
+	@Autowired
+	private ObjectMapper mapper;
+
+	@Autowired
+	private PayService payService;
+
+	@Autowired
+	private MasterDataService mstrDataService;
+
+	@Autowired
+	private WSCalculationUtil utils;
+
+	@Autowired
+	private WSCalculationConfiguration configs;
+
+	@Autowired
+	private ServiceRequestRepository serviceRequestRepository;
+
+	@Autowired
+	private DemandRepository demandRepository;
+	
+    @Autowired
+    WSCalculationService wsCalculationService;
+    
+    @Autowired
+    WSCalculationDao waterCalculatorDao;
+    
+    @Autowired
+    CalculatorUtil calculatorUtils;
+
+
+	/**
+	 * Creates or updates Demand
+	 * 
+	 * @param requestInfo
+	 *            The RequestInfo of the calculation request
+	 * @param calculations
+	 *            The Calculation Objects for which demand has to be generated
+	 *            or updated
+	 */
+	public List<Demand> generateDemand(RequestInfo requestInfo, List<Calculation> calculations,
+			Map<String, Object> masterMap) {
+		List<Demand> createdDemands = new ArrayList<>();
+		// List that will contain Calculation for new demands
+		List<Calculation> createCalculations = new LinkedList<>();
+
+		
+		@SuppressWarnings("unchecked")
+		Map<String, Object> financialYearMaster =  (Map<String, Object>) masterMap
+				.get(WSCalculationConstant.BillingPeriod);
+		Long fromDate = (Long) financialYearMaster.get(WSCalculationConstant.STARTING_DATE_APPLICABLES);
+		Long toDate = (Long) financialYearMaster.get(WSCalculationConstant.ENDING_DATE_APPLICABLES);
+		
+		// List that will contain Calculation for old demands
+		List<Calculation> updateCalculations = new LinkedList<>();
+
+		if (!CollectionUtils.isEmpty(calculations)) {
+			// Collect required parameters for demand search
+			String tenantId = calculations.get(0).getTenantId();
+			Set<String> consumerCodes = calculations.stream().map(calculation -> calculation.getConnectionNo())
+					.collect(Collectors.toSet());
+			List<Demand> demands = searchDemand(tenantId, consumerCodes, fromDate, toDate, requestInfo);
+			Set<String> connectionNumbersFromDemands = new HashSet<>();
+			if (!CollectionUtils.isEmpty(demands))
+				connectionNumbersFromDemands = demands.stream().map(Demand::getConsumerCode)
+						.collect(Collectors.toSet());
+
+			// If demand already exists add it updateCalculations else
+			// createCalculations
+			for (Calculation calculation : calculations) {
+				if (!connectionNumbersFromDemands.contains(calculation.getConnectionNo()))
+					createCalculations.add(calculation);
+				else
+					updateCalculations.add(calculation);
+			}
+		}
+
+		if (!CollectionUtils.isEmpty(createCalculations))
+			createdDemands = createDemand(requestInfo, createCalculations, masterMap);
+
+		if (!CollectionUtils.isEmpty(updateCalculations))
+			createdDemands = updateDemandForCalculation(requestInfo, updateCalculations, fromDate, toDate);
```

Comment:
CreatedDemands variable is being overwritten

Reply:
here if demand already exists only then updatedemandforcalculatio will be called else create demand will be called